src/devtools/client/webconsole/components/Output/ConsoleOutput.tsx

```diff
@@ -202,12 +222,125 @@ class ConsoleOutput extends React.Component<PropsFromRedux> {
     });
 
     return (
-      <div className="webconsole-output" ref={this.outputNode} role="main">
-        <ConsoleLoadingBar />
-        {messageNodes}
-      </div>
+      <>
+        <div
+          className="webconsole-output"
+          ref={this.outputNode}
+          role="main"
+          onContextMenu={this.onContextMenu}
+        >
+          <ConsoleLoadingBar />
+          {messageNodes}
+        </div>
+        {contextMenu !== null && (
+          <ContextMenu x={contextMenu.pageX} y={contextMenu.pageY} close={this.closeContextMenu}>
+            <Dropdown>
+              <DropdownItem onClick={this.setFocusStart}>
+                <>
+                  <Icon filename="set-focus-start" className="mr-4 bg-iconColor" />
+                  Set focus start
+                </>
+              </DropdownItem>
+              <DropdownItem onClick={this.setFocusEnd}>
+                <>
+                  <Icon filename="set-focus-end" className="mr-4 bg-iconColor" />
+                  Set focus end
+                </>
+              </DropdownItem>
+            </Dropdown>
+          </ContextMenu>
+        )}
+      </>
     );
   }
+
+  closeContextMenu = () => {
+    this.setState({ contextMenu: null });
+  };
+
+  onContextMenu = (event: React.MouseEvent) => {
+    // This method of mapping context-menu event to a message is a bit hacky.
+    // This section of the UI was written in a pretty imperative style though.
+    // Ideally we will revisit and refactor the console UI components at some point.
+    let message: Message | null = null;
+    let target: HTMLElement | null = event.target as HTMLElement;
+    while (target) {
+      if (target.hasAttribute("data-message-id")) {
+        const id = parseInt(target.getAttribute("data-message-id") as string);
+        message = this.props.messages.entities[id] || null;
+        break;
+      }
+
+      target = target.parentElement;
+    }
+
+    if (message == null) {
+      return;
+    }
+
+    event.stopPropagation();
+    event.preventDefault();
+
+    this.setState({
+      contextMenu: {
+        message,
+        pageX: event.pageX,
+        pageY: event.pageY,
+      },
+    });
+  };
+
+  setFocusEnd = async () => {
+    const { dispatch, focusRegion } = this.props;
+    const { message } = this.state.contextMenu!;
+
+    this.setState({ contextMenu: null });
+
+    const time = await getTimeForMessage(message);
+    if (time < 0) {
+      console.error("Could not calculate time for message", message);
+      return;
+    }
+
+    // If this is the first time the user is focusing, start at the beginning of the recording (or zoom region).
+    // Let the focus action/reducer will handle cropping for us.
+    const startTime = focusRegion ? focusRegion.startTime : 0;
+    const endTime = time!;
+
+    dispatch(
+      setFocusRegion({
+        endTime,
+        startTime,
+      }) as unknown as AnyAction
+    );
+    dispatch(syncFocusedRegion() as unknown as AnyAction);
```

Comment:
cc @markerikson This is a silly type refinement. Any suggestions?

Reply:
Yeah, the issue here is that the base `Dispatch` type from Redux knows nothing about thunks, and `syncFocusedRegion` is a thunk.

On top of that, `connect` by itself only knows about the base `Dispatch` type.  Same with `useDispatch`, but it's a lot easier to create a custom hook that uses the _real_ type inferred from `store.dispatch`: https://redux.js.org/tutorials/typescript-quick-start#define-typed-hooks

The right answer here is to do `import type { AppDispatch } from "ui/setup/store";`, which has the correct types for the thunk middleware baked in because that was part of the store setup.  Then we need to re-type `dispatch` inside of this block as an `AppDispatch` instead.  (Not sure if there's a good way to do it at the `connect` level, unless we were to manually provide a `mapDispatch` function and cast in there).

That will make it safe to do `dispatch(syncFocusedRegion())` without actually casting on that line.